diff -u -r qt-everywhere-opensource-src-4.7.1.orig/mkspecs/common/g++.conf qt-everywhere-opensource-src-4.7.1/mkspecs/common/g++.conf
--- qt-everywhere-opensource-src-4.7.1.orig/mkspecs/common/g++.conf	2010-11-05 18:55:11.000000000 -0700
+++ qt-everywhere-opensource-src-4.7.1/mkspecs/common/g++.conf	2010-12-21 13:59:03.000000000 -0800
@@ -17,7 +17,7 @@
 QMAKE_CFLAGS_PRECOMPILE += -x c-header -c ${QMAKE_PCH_INPUT} -o ${QMAKE_PCH_OUTPUT}
 QMAKE_CFLAGS_USE_PRECOMPILE += -include ${QMAKE_PCH_OUTPUT_BASE}
 
-QMAKE_CXX		= g++
+QMAKE_CXX		= g++-4.1
 QMAKE_CXXFLAGS		+= $$QMAKE_CFLAGS
 QMAKE_CXXFLAGS_DEPS	+= $$QMAKE_CFLAGS_DEPS
 QMAKE_CXXFLAGS_WARN_ON	+= $$QMAKE_CFLAGS_WARN_ON
@@ -32,8 +32,8 @@
 QMAKE_CXXFLAGS_PRECOMPILE += -x c++-header -c ${QMAKE_PCH_INPUT} -o ${QMAKE_PCH_OUTPUT}
 QMAKE_CXXFLAGS_USE_PRECOMPILE = $$QMAKE_CFLAGS_USE_PRECOMPILE
 
-QMAKE_LINK		= g++
-QMAKE_LINK_SHLIB	= g++
+QMAKE_LINK		= g++-4.1
+QMAKE_LINK_SHLIB	= g++-4.1
 QMAKE_LINK_C		= gcc
 QMAKE_LINK_C_SHLIB	= gcc
 QMAKE_LFLAGS		+=
