From 54ba3533c79b860758c2839e71d5771e59a5a889 Mon Sep 17 00:00:00 2001
From: Benjamin C Meyer <benjamin.meyer@torchmobile.com>
Date: Mon, 23 Mar 2009 17:58:00 -0400
Subject: [PATCH] Disable dragging in qtwebkit

---
 .../WebKit/qt/WebCoreSupport/DragClientQt.cpp      |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/src/3rdparty/webkit/WebKit/qt/WebCoreSupport/DragClientQt.cpp b/src/3rdparty/webkit/WebKit/qt/WebCoreSupport/DragClientQt.cpp
index b719868..abd5a00 100644
--- a/src/3rdparty/webkit/WebKit/qt/WebCoreSupport/DragClientQt.cpp
+++ b/src/3rdparty/webkit/WebKit/qt/WebCoreSupport/DragClientQt.cpp
@@ -59,6 +59,7 @@ void DragClientQt::willPerformDragSourceAction(DragSourceAction, const IntPoint&
 
 void DragClientQt::startDrag(DragImageRef, const IntPoint&, const IntPoint&, Clipboard* clipboard, Frame*, bool)
 {
+    return;
 #ifndef QT_NO_DRAGANDDROP
     QMimeData* clipboardData = static_cast<ClipboardQt*>(clipboard)->clipboardData();
     static_cast<ClipboardQt*>(clipboard)->invalidateWritableData();
-- 
1.6.0.4

