*** Source: http://marc.info/?l=openssl-dev&m=107628900919852&q=raw
*** openssl-0.9.7c/util/mk1mf.pl	Fri Mar 14 17:29:18 2003
--- openssl-0.9.7c-miket/util/mk1mf.pl	Sat Feb  7 21:58:57 2004
***************
*** 485,490 ****
--- 485,495 ----
  	chop;
  
  	($key,$val)=/^([^=]+)=(.*)/;
+ 	
+ 	# On some Windows machines, $val has linefeeds at the end, which confuses
+ 	# subsequent code in this file. So we strip all whitespace at the end.
+     $val =~ s/\s+$//;
+     
  	if ($key eq "RELATIVE_DIRECTORY")
  		{
  		if ($lib ne "")
